test/demo/accounts_test.exs

```diff
@@ -0,0 +1,471 @@
+
+    test "returns the user if the email exists" do
+      %{id: id} = user = user_fixture()
+      assert %User{id: ^id} = Accounts.get_user_by_email(user.email)
+    end
+  end
+
+  describe "get_user_by_email_and_password/1" do
+    test "does not return the user if the email does not exist" do
+      refute Accounts.get_user_by_email_and_password("unknown@example.com", "hello world!")
+    end
+
+    test "does not return the user if the password is not valid" do
+      user = user_fixture()
+      refute Accounts.get_user_by_email_and_password(user.email, "invalid")
+    end
+
+    test "returns the user if the email and password are valid" do
+      %{id: id} = user = user_fixture()
+
+      assert %User{id: ^id} =
+               Accounts.get_user_by_email_and_password(user.email, valid_user_password())
+    end
+  end
+
+  describe "get_user!/1" do
+    test "raises if id is invalid" do
+      assert_raise Ecto.NoResultsError, fn ->
+        Accounts.get_user!(123)
+      end
+    end
+
+    test "returns the user with the given id" do
+      %{id: id} = user = user_fixture()
+      assert %User{id: ^id} = Accounts.get_user!(user.id)
+    end
+  end
+
+  describe "register_user/1" do
+    test "requires email and password to be set" do
+      {:error, changeset} = Accounts.register_user(%{})
+
+      assert %{
+               password: ["can't be blank"],
+               email: ["can't be blank"]
+             } = errors_on(changeset)
+    end
+
+    test "validates email and password when given" do
+      {:error, changeset} = Accounts.register_user(%{email: "not valid", password: "not valid"})
+
+      assert %{
+               email: ["must have the @ sign and no spaces"],
+               password: ["should be at least 12 character(s)"]
+             } = errors_on(changeset)
+    end
+
+    test "validates maximum values for e-mail and password for security" do
+      too_long = String.duplicate("db", 100)
+      {:error, changeset} = Accounts.register_user(%{email: too_long, password: too_long})
+      assert "should be at most 160 character(s)" in errors_on(changeset).email
+      assert "should be at most 80 character(s)" in errors_on(changeset).password
+    end
+
+    test "validates e-mail uniqueness" do
+      %{email: email} = user_fixture()
+      {:error, changeset} = Accounts.register_user(%{email: email})
+      assert "has already been taken" in errors_on(changeset).email
+
+      # Now try with the upper cased e-mail too, to check that email case is ignored.
+      {:error, changeset} = Accounts.register_user(%{email: String.upcase(email)})
+      assert "has already been taken" in errors_on(changeset).email
+    end
+
+    test "registers users with a hashed password" do
+      email = unique_user_email()
+      {:ok, user} = Accounts.register_user(%{email: email, password: valid_user_password()})
+      assert user.email == email
+      assert is_binary(user.hashed_password)
+      assert is_nil(user.confirmed_at)
```

Comment:
Should we also assert that `user.password` is `nil`? I'm thinking that there could be a security potential security issue of the user returned by `register_user/1` returns a user with a plain text password. 

If we do this, I think a similar update will need be made for `update_user_password/2`.

Reply:
I wonder if adding `@derive {Inspect, except: :password}` is enough? Returning the user with a password will only be a problem if you explicitly write it down somewhere and filtering the inspect will help avoid most cases of that.

Reply:
I originally found this issue when writing code like this for a user profile form (that allows updating name, email, and password all in the same form).

```elixir
# Test is adapted from the code generated by phx.gen.context
test "update_user/2 with invalid data returns error changeset" do
  user = user_fixture()
  invalid_attrs = %{email: "invalid"}
  assert {:error, changeset} = Accounts.update_user(user, invalid_attrs)
  assert ^user = Accounts.get_user!(user.id)
end
```

This code doesn't work because the user coming back from `user_fixture/0` (which calls `Accounts.register_user/1`) returns a user with the password field filled in and `Accounts.get_user!/1` returns password as `nil`.

My fix in my app was to add a call to `Changeset.delete_change/2` in `Accounts.User.maybe_hash_password/1`

```elixir
  defp maybe_hash_password(changeset) do
    password = get_change(changeset, :password)

    if password && changeset.valid? do
      changeset
      |> put_change(:hashed_password, Bcrypt.hash_pwd_salt(password))
      |> delete_change(:password)
    else
      changeset
    end
  end
```

I think `@derive {Inspect, except: :password}` is a great idea, but the problem still remains that a user with a non-nil password may be passed around the system unexpectedly if it was returned from `Accounts.register_user/1` or  `Accounts.update_user_password/2`. If the user decided to do some sort of serialization using `Map.from_struct/1` there's a chance they could accidentally serialize a plain text password when the password is normally `nil` (not that serializing a user struct this way is a good practice, but I'm sure it happens).

Reply:
Done!